### DISCUSSION TITLE
Risks of not using `rangeStrategy: 'bump'`

### DISCUSSION THREAD

**Discussion by simonhaenisch** [renovatebot/renovate#23275] [Suggest an Idea]
### Tell us more.

I recently discussed this at work and there's this thing about semver and ranges that I ended up wondering whether it has been considered in the `auto` range strategy (at least for npm). It's related to the current behavior as stated in the `rangeStrategy` docs:

> By default, Renovate assumes that if you are using ranges then it's because you want them to be wide/open. Renovate won't deliberately "narrow" any range by increasing the semver value inside.
>
> For example, if your package.json specifies a value for left-pad of ^1.0.0 and the latest version on npmjs is 1.2.0, then Renovate won't change anything because 1.2.0 satisfies the range. If instead you'd prefer to be updated to ^1.2.0 in cases like this, then configure rangeStrategy to bump in your Renovate config.

There's the following case where only bumping the lock file but not the `package.json` could have an impact:

- I use a function of package `foo` version `1.2.0` in my library.
- A new version `1.3.0` is released which introduces a new feature, e.g. adds a new function.
- Renovate updates me to the latest version.
- I now use the new feature in my library and publish a new version.
- End users update my library but it doesn't automatically bump the transitive dependency to `foo` because their currently locked `1.2.0` already satisfies my range.
- Now their code breaks.

The `bump` range strategy solves this problem by forcing my library users to upgrade the transitive dependency as well. Wondering whether this problem should be considered in the `auto` strategy, or it makes sense to make `bump` the default at some point?

**Comment by rarkins** [comment]
You may as well use pinned versions in that case

**Reply by simonhaenisch** [reply]
I don't really follow... using a pinned version would mean I don't allow them to upgrade the transitive dependency to a newer compatible one.

What I mean is that by not bumping the range in package.json along with an update, I could be using a feature introduced in a newer version (semver minor) than the minimum one defined by the range. It would work in my project because of my lock file with the newest version. But only the package.json gets published with a technically invalid range. Fine when resolving for the first time, but at least not all package managers seem to upgrade the respective transitive dependencies when upgrading a dependency.

**Reply by rarkins** [reply]
I understand what you're describing completely. What you're describing is "dangers of supporting more than one version of a dependency without testing every version". It's a valid concern but is not caused by Renovate.

It's common that a library author supports dependency ranges like `^1.0.0`. It's also common that library authors only test on the latest version, e.g. 1.5.0. This risk happens whether you use Renovate or not. Keeping a 1-2 year old lock file is not the solution to this.

**Reply by simonhaenisch** [reply]
> This risk happens whether you use Renovate or not.

Before Renovate, I used to use e.g. `npm install dep-name@latest` which would also bump the range in `package.json`. Meaning if I published a new version with a new feature relying on that library, I'd force my library users to update the transitive dependency to a new minimum. Renovate can do this with `bump` range strategy, so I just wanted to discuss if it's a less risky default/auto strategy.

> Keeping a 1-2 year old lock file is not the solution to this.

Not sure what you mean by this exactly... doing range bumps in `package.json` actually means I force my library users to upgrade the transitive dependencies in their lockfile to newer versions, right?

Anyway, we can close this discussion, I was just interested to hear your thoughts (:

**Comment by rarkins** [comment]
> > This risk happens whether you use Renovate or not.
> 
> Before Renovate, I used to use e.g. `npm install dep-name@latest` which would also bump the range in `package.json`. Meaning if I published a new version with a new feature relying on that library, I'd force my library users to update the transitive dependency to a new minimum. Renovate can do this with `bump` range strategy, so I just wanted to discuss if it's a less risky default/auto strategy.

Your paragraph above implies that you _know_ that you relied on a new feature of that library. Yet earlier you are basing your "danger" on the developer not being aware.

Ultimately, users want a library to depend on as wide of a constraint as possible, e.g. `^1.0.0`. If at any time a developer uses a feature released `>1.0.0`, that constraint is no longer valid. If they know they are using a new feature, then there is no problem - they can update their dependency constraint regardless of whether they are using Renovate or not. If they don't realize what they've done, they'll also have a problem regardless of using Renovate's default behavior or not. The only way to avoid such human error is to use a pinned version, but then you are not meeting the user's desires to support a wide range.

BTW one way library authors could overcome this is to test against not just the latest/newest version of each dependency, but also the oldest. If you test both ends of the range, you are likely ok.

**Reply by simonhaenisch** [reply]
I would often (manually) update dependencies as a chore before starting to work on a project again, but I would not check for all their new features. Intellisense or the likes would just suggest me any library methods, including new ones. So I wouldn't say I'd always be aware that I'm relying on a feature of the updated version.

As I said with my manual updates I would also bump the range so this risk doesn't exist. With Renovate continuously updating my project, when I jump back into working on a project after a while, I'll just work with whatever versions there currently are but the allowed ranges haven't changed, so this becomes more of a risk imo.

Anyway, yes testing with both locked version and range start would solve this problem. It's an interesting idea actually, also for peer dependencies.

Also, the chances of this whole thing happening are quite slim, plus it's easy enough to forward fix my library with a new patch version and deprecating the faulty one.
